File: project-euler.go/solutions/_4/4.go
package _4

func Solve() int64 {
	leftSide := 999
	num      := makePalindrome(leftSide)

	for !check(num) {
		leftSide--
		num = makePalindrome(leftSide)
	}

	return int64(num)
}

func makePalindrome(num int) int {
	c := num % 10
	b := (num / 10) % 10
	a := num / 100

	return num * 1000 + c * 100 + b * 10 + a
}

func check(num int) bool {
	if num % 11 != 0  {
		return false
	}

	for fac := 999; fac >= 900; fac-- {
		if num % fac == 0 && num / fac < 1000 {
			return true
		}
	}

	return false
}

File: project-euler.java/src/main/java/euler/solutions/_4.java
package euler.solutions;

import euler.common.BaseSolution;

public class _4 extends BaseSolution {

    @Override
    public long solve() {
        int leftSide = 999;
        int num = makePalindrome(leftSide);

        while (!check(num)) {
            leftSide--;
            num = makePalindrome(leftSide);
        }

        return num;
    }

    int makePalindrome(int num) {
        int c = num % 10;
        int b = (num / 10) % 10;
        int a = num / 100;

        return num * 1000 + c * 100 + b * 10 + a;
    }

    boolean check(int num) {
        if (num % 11 != 0) return false;

        for (int fac = 999; fac >= 900; fac--) {
            if (num % fac == 0 && num / fac < 1000) {
                return true;
            }
        }

        return false;
    }
}
